distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso

```diff
@@ -2706,3 +2836,7 @@ Table.from (that:Materialized_Table) =
 
 ## PRIVATE
 Table_Ref.from (that:Table) = Table_Ref.Value that
+
+## PRIVATE
+    The largest dataset that can be used to make a literal table, expressed in number of elements.
+MAX_LITERAL_ELEMENT_COUNT = 256
```

Comment:
Perhaps this should be bigger?

Reply:
Is there any DB limit? I guess it is limited by the query size... I wonder if such literal table will be able to use hashjoins or will it always fall back to a linear scan. I guess for <256 it doesn't matter much. For larger values the size of the SQL query may start being quite problematic.

I think ideally we shouldn't cut off but instead we should be creating temporary tables - but definitely an improvement for a separate PR.
(I also wonder, given how often currently Enso re-evaluates the expressions, if we could create too much garbage in such way)

Reply:
Yes, that was my concern. My first thought was to create a temporary table, but that will be happening repeatedly during each evaluation.

I was able to create tables with ~18000 rows this way in both postgres and sqlite, and at that point I stopped testing. I'm more concerned about the size of the query being sent over the wire, but I don't really know at what point that becomes a problem. I figured that giving it a small would be a good first step. I think it covers the majority of use-cases.

Reply:
Interesting that it worked for such large sizes.

I guess the primary concern indeed is the query size, especially as a temp table is likely sent in a more compact binary compact.

Anyway, seems all ok for now.

Reply:
256 a reasonable literal default. Agee with @radeusgd should be a temp table when gets bigger!

Reply:
Will this result in a lot of temporary tables being created and not eagerly cleaned up? That was my concern when doing it the literal way.

Reply:
> Will this result in a lot of temporary tables being created and not eagerly cleaned up? That was my concern when doing it the literal way.

Yes indeed that is a concern.

I think we should implement such a feature as a separate PR, as it has non-trivial complexity.

Also, I think the concern of 'too many' temporary tables may also be increased due to the fact that currently the IDE seems to recompute unrelated nodes 'too often'. I think that if we resolve that, the amount of re-computations could be lower and thus the problem would not be as big.

As for implementing these temporary tables efficiently, we can leverage a few tricks:
1. I expect that very often when the operation is re-run, the in-memory lookup table will actually be the same between re-runs. We can exploit that and try keeping a cache of already uploaded temporary tables, indexed by the hashcode of the table's contents (uploading the table requires us to scan its whole contents anyway, so the additional cost of computing the hashcode is negligible in this case). This way we can avoid uploading a _new_ temporary table on each run, if we can detect that a 'matching' table was already uploaded before.
2. We can exploit the `Managed_Resource` framework to try to clean up the tables once the references to them are GCed. We actually already implement a very similar feature - `Hidden_Table_Registry`. It is used to be able to re-use temporary hidden tables for dry-run operations, and clean them up once they are no longer needed. We could extend this registry to also support such temporary tables that are not used for dry-run (and accessed by name) but accessed by e.g. content's hash.

In fact we can merge the 2 approaches. If between runs the tables are the same we can avoid re-uploading them. Once all references to them are GCed we can clean them up.

Reply:
https://github.com/enso-org/enso/issues/9086